### PR TITLE
Exploring the possibility of a ChainHandler

### PR DESCRIPTION
**Pull request by theninj4**
This is a work in progress and a proof of concept for a possible solution to #176 the idea being we can chain together handlers:

### Chaining handlers together with the ChainHandler

This project ships with a handler called the `ChainHandler`. It's purpose is to make it easier to wrap the functionality of one handler with additional functionality. For example using the relational data store may be ideal for storing resources, but what if we want to add an authentication layer on top of those resources? An authentication handler shouldn't know or care about the storage of the underlying resource, it should only deal with authentication - we need an abstraction layer to help us glue all of this together. 

To get started with the ChainHandler, go ahead and instantiate one:
```javascript
var jsonApi = require("jsonapi-server");
var testChainHandler = new jsonApi.ChainHandler();
```
From there it's a simple case of attaching new function handlers to `testChainHandler`, following the naming convention of `("before"|"after")+function` - eg `beforeSearch`, `beforeCreate`, `afterDelete`.

These new functions need to match the signatures of the functions they're overriding and need to pass the data onwards via a callback. For example, here are two valid functions that don't add any functionality. Notice how the callback values match the function signatures:
```javascript
testChainHandler.beforeSearch = function(request, callback) {
  return callback(null, request);
};

testChainHandler.afterSearch = function(results, pagination, callback) {
  return callback(null, results, pagination);
};
```

With a new ChainHandler ready to go, this is how we can stack them up:
```javascript
jsonApi.define({
  resource: "articles",
  handlers: chainHandler1.chain(chainHandler2).chain(memoryHandler),
});
```

The call stack when chaining multiple `ChainHandler`s together looks like this:
```
chainHandler1.beforeSearch
  chainHandler2.beforeSearch
    MemoryHandler.search
  chainHandler2.afterSearch
chainHandler1.afterSearch
```

### CODE REVIEW

File: lib/ChainHandler.js [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+
+var ChainHandler = module.exports = function ChainHandler() {
+};
+
+ChainHandler.prototype.chain = function(otherHandler) {
+  var self = this;
+  self.otherHandler = otherHandler;
+  self.ready = true
+  return self;
+};
```
Comment (pmcnr-hx) [inline lib/ChainHandler.js]: Couldn't we just move the `chain` functionality to the constructor function?

File: README.md [hunk 2]
```diff
@@ -34,6 +34,7 @@ Ultimately, the only things you as a user of this framework need to care about a
 
 We've created `handler`s to automatically map our config over to database solutions help people get off the ground:
  * [jsonapi-store-memoryhandler](https://github.com/holidayextras/jsonapi-server/blob/master/lib/MemoryHandler.js) - an in-memory data store to enable rapid prototyping. This ships as a part of `jsonapi-server` and powers the core test suite.
+ * [jsonapi-store-chainhandler](https://github.com/holidayextras/jsonapi-server/blob/master/lib/ChainHandler.js) - a handler to chain custom behaviour around an existing handler. This ships as a part of `jsonapi-server`.
```
Comment (pmcnr-hx) [inline README.md]: `jsonapi-handler-chain` instead, since it's not really a store?

**Comment by pmcnr-hx** [comment]
This is marvellous! (ignore my outdated comment, having the separate `.chain` method reads much better).

File: README.md [hunk 3]
```diff
@@ -34,6 +34,7 @@ Ultimately, the only things you as a user of this framework need to care about a
 
 We've created `handler`s to automatically map our config over to database solutions help people get off the ground:
  * [jsonapi-store-memoryhandler](https://github.com/holidayextras/jsonapi-server/blob/master/lib/MemoryHandler.js) - an in-memory data store to enable rapid prototyping. This ships as a part of `jsonapi-server` and powers the core test suite.
+ * [jsonapi-handler-chain](https://github.com/holidayextras/jsonapi-server/blob/master/lib/ChainHandler.js) - a handler to chain custom behaviour around an existing handler. This ships as a part of `jsonapi-server`. More info can be found [here](https://github.com/holidayextras/jsonapi-server/blob/master/documentation/chain-handler.js)
```
Comment (pmcnr-hx) [inline README.md]: `...documentation/chain-handler.js` → `...documentation/chain-handler.md`

File: lib/ChainHandler.js [hunk 4]
```diff
@@ -0,0 +1,46 @@
+/* @flow weak */
+"use strict";
+
+var async = require("async");
+
+var ChainHandler = module.exports = function ChainHandler() { };
+
+ChainHandler.prototype.chain = function(otherHandler) {
```
Comment (pmcnr-hx) [inline lib/ChainHandler.js]: I'm wondering if calling the parameter `nextHandler` would make clearer the the order of invocation... Just a suggestion though, it's fine as is.

**Comment by pmcnr-hx** [comment]
Excellent work, and all looks legit! Worth adding a few automated tests to cover the new functionality and avoid future regressions?

**Comment by theninj4** [comment]
I've left a dual-chained handler in the example project used by the tests - if we break the ChainHandler the test suite fails catastrophically ;)

File: lib/ChainHandler.js (hunk 4 above)
Comment (theninj4) [inline lib/ChainHandler.js]: I like the use of `chain` as when we start stacking up these handlers it won't be very obvious where the magic is coming from - by including a part of the name in the function we're helping to steer newcomers towards the right documentation.

**Pull request merged**